lib/hex_web/validation.ex

```diff
@@ -52,27 +17,14 @@ defmodule HexWeb.Validation do
     end)
   end
 
-  def validate_meta(changeset, field, meta_types, meta_required) do
-    validate_change(changeset, field, fn _field, meta ->
-      type_errors =
-        Enum.flat_map(meta_types, fn {sub_field, type} ->
-          type(sub_field, Map.get(meta, sub_field), type)
-        end)
-
-      req_errors =
-        Enum.flat_map(meta_required, fn field ->
-          if Map.has_key?(meta, field) do
-            []
-          else
-            [{field, :missing}]
-          end
-        end)
-
-      errors = req_errors ++ type_errors
-
-      if errors == [],
-          do: [],
-        else: [{field, errors}]
-    end)
+  @doc """
+  Adds embed errors to the changeset.errors
+  """
+  def put_embed_errors(%Ecto.Changeset{changes: changes} = changeset, embed) do
```

Comment:
yeah, I already started looking into `traverse_errors` to be used by
`validation_failed(conn, errors)` helper internally, but I also noticed
there's some existing code that expected `errors[:meta]` to be a list, e.g.
here:
https://github.com/hexpm/hex_web/pull/258/files#diff-c3cdcf0a302e924598045042d82d28cdR42.
Do you have a preference how this should be handled? I could leave this
`put_embed_error` helper, or I could use traverse everywhere including
tests, same as in
https://github.com/elixir-lang/ecto/blob/v2.0.0-beta.1/test/ecto/changeset/embedded_test.exs#L705

2016-03-06 10:11 GMT+01:00 José Valim <notifications@github.com>:

> In lib/hex_web/validation.ex
> <https://github.com/hexpm/hex_web/pull/258#discussion_r55135565>:
>
> > -            []
> > -          else
> > -            [{field, :missing}]
> > -          end
> > -        end)
> > -
> > -      errors = req_errors ++ type_errors
> > -
> > -      if errors == [],
> > -          do: [],
> > -        else: [{field, errors}]
> > -    end)
> > +  @doc """
> > +  Adds embed errors to the changeset.errors
> > +  """
> > +  def put_embed_errors(%Ecto.Changeset{changes: changes} = changeset, embed) do
>
> The idea is to keep all errors in the changesets and use the
> Ecto.Changeset.traverse_errors functions to traverse them into a single
> list.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hexpm/hex_web/pull/258/files#r55135565>.
>



-- 
Wojtek Mach